### PR TITLE
AB#4378 drop target if unequal

### PR DESCRIPTION
**Pull request by jjmurre**
See commit message for explanation.

### CODE REVIEW

File: src/plugins/postgres_table_copy_operator.py [hunk 1]
```diff
@@ -112,6 +116,50 @@ def __init__(
...
+               WHERE table_name = %(source_table_name)s)
+            """
+                ),
+                (
+                    {
+                        "source_table_name": table_mapping.source,
+                        "target_table_name": table_mapping.target,
+                    }
+                ),
+            )
```
Comment (lbam) [inline src/plugins/postgres_table_copy_operator.py:154]: Why the extra parentheses?

File: src/plugins/postgres_table_copy_operator.py [hunk 2]
```diff
@@ -112,6 +116,50 @@ def __init__(
...
+                      ordinal_position,
+                      column_name
+               FROM information_schema.columns
+               WHERE table_name = %(target_table_name)s
+               EXCEPT SELECT data_type,
+                             ordinal_position,
+                             column_name
+               FROM information_schema.columns
+               WHERE table_name = %(source_table_name)s)
+            """
```
Comment (lbam) [inline src/plugins/postgres_table_copy_operator.py:146]: You can do this in less code with a full outer join:

```sql
WITH src AS (
    SELECT data_type, ordinal_position, column_name FROM information_schema.columns
    WHERE table_name = %(source_table_name)s
), tgt AS (
    SELECT data_type, ordinal_position, column_name FROM information_schema.columns
    WHERE table_name = %(source_table_name)s
)
SELECT (COUNT(*) = COUNT(src)) FROM src NATURAL FULL OUTER JOIN tgt
```

File: src/plugins/postgres_table_copy_operator.py (hunk 1 above)
Comment (jjmurre) [reply src/plugins/postgres_table_copy_operator.py:154]: Ah, that was a left-over from editing stuff, removed! Tx.

File: src/plugins/postgres_table_copy_operator.py (hunk 2 above)
Comment (jjmurre) [reply src/plugins/postgres_table_copy_operator.py:146]: That's a nice improvement, I'll replace the query with this one.

Comment (lbam) [reply src/plugins/postgres_table_copy_operator.py:146]: Please test it first, I didn't test this final version with the string interpolation.

Comment (jjmurre) [reply src/plugins/postgres_table_copy_operator.py:146]: Yes, I tested it. There turned out the be one flow. If the target table was smaller, it did not work, so I changed the condition in the select to:

                        SELECT COUNT(*) = COUNT(src) AND count(*) = COUNT(tgt)

Comment (jjmurre) [reply src/plugins/postgres_table_copy_operator.py:146]: And during testing, I discovered a bug. There was no check on the existence of the target table, so, I added an extra check.

Comment (lbam) [reply src/plugins/postgres_table_copy_operator.py:153]: Oh, good catch.

File: src/plugins/postgres_table_copy_operator.py [hunk 3]
```diff
@@ -119,40 +120,44 @@ def __init__(
     def _drop_tables_if_unequal(self, cursor: Any, all_table_copies: List[TableMapping]) -> None:
         """Drop the target tables that are unequal to their source table."""
         for table_mapping in all_table_copies:
+            # First make sure the target table exists
+            cursor.execute(
+                sql.SQL("SELECT to_regclass('{target_table_name}')").format(
+                    target_table_name=sql.Identifier(table_mapping.target)
+                )
+            )
```
Comment (lbam) [inline src/plugins/postgres_table_copy_operator.py:128]: Are you sure this is necessary? When I run the outer join query with a non-existing table name, I get a false result (assuming the other table exists).

**Review by lbam — APPROVED** [review]

File: src/plugins/postgres_table_copy_operator.py (hunk 3 above)
Comment (jjmurre) [reply src/plugins/postgres_table_copy_operator.py:128]: The problem is, false result is interpreted as tables are not equal. So next step is to remove the target table. However, that does not exist, leading to crash in postgres.

Comment (lbam) [reply src/plugins/postgres_table_copy_operator.py:128]: Oh, right. Ok!

File: src/plugins/postgres_table_copy_operator.py [hunk 4]
```diff
@@ -112,6 +117,54 @@ def __init__(
                 "even though source table will be dropped."
             )
 
+    def _drop_tables_if_unequal(self, cursor: Any, all_table_copies: List[TableMapping]) -> None:
+        """Drop the target tables that are unequal to their source table."""
+        for table_mapping in all_table_copies:
+            # First make sure the target table exists
+            cursor.execute(
+                sql.SQL("SELECT to_regclass('{target_table_name}')").format(
```
Comment (gkoller) [inline src/plugins/postgres_table_copy_operator.py:125]: I'm very glad to see `psycopg2.sql` used in more and more code in favor of `str` formatting.

Here however a regular DBAPI parameter passing would have sufficed. Eg:

```python
cursor.execute(
    "SELECT to_regclass(%(target_table_name)s)",
    {target_table_name=table_mapping.target}
)
```

Reason being that `target_table_name` is not really an identifier from the SQL grammar point of view. Yes, it is a table name and table names are identifiers, but here it is just a string argument to a PostgreSQL function. Simply using `%(target_table_name)s` will make sure it is quoted as a string argument (hence you can leave the quotes out).

And as we're just using a regular SQL query with plain DBAPI parameters, you can do away with `sql.SQL` and `sql.Identifier`

File: src/plugins/postgres_table_copy_operator.py [hunk 5]
```diff
@@ -112,6 +117,54 @@ def __init__(
...
+                )
+            )
+
+            if first(cursor.fetchone()) is None:
+                self.log.info(
+                    "Target table %s not found, skipping equality check.", table_mapping.target
+                )
+                continue
+            cursor.execute(
+                sql.SQL(
```
Comment (gkoller) [inline src/plugins/postgres_table_copy_operator.py:136]: No need for `sql.SQL` here. You are not doing anything fance with `sql.Identifier` or `sql.Placeholder` or whatever. This is just a regular SQL query with DBAPI parameters (the `%(...)s ones).

BTW I like this query. Nice solution.

File: src/plugins/postgres_table_copy_operator.py [hunk 6]
```diff
@@ -112,6 +117,54 @@ def __init__(
...
+                {
+                    "source_table_name": table_mapping.source,
+                    "target_table_name": table_mapping.target,
+                },
+            )
+            if not first(cursor.fetchone()):
+                self.log.info("Dropping table %s", table_mapping.target)
+                cursor.execute(
+                    sql.SQL("DROP TABLE {table_name}").format(
+                        table_name=sql.Identifier(table_mapping.target)
```
Comment (gkoller) [inline src/plugins/postgres_table_copy_operator.py:164]: Yes! :+1:  Here the table name is indeed an identifier from the SQL grammar point of view. Hence you need to use `postgresql.sql` helpers.

**Review by gkoller — APPROVED** [review]
Very nice addition. Adding this here allows us to remove many `rm_tmp_table` tasks in our DAGs :+1:

**Pull request merged**